pkg/collect/etcd.go

```diff
@@ -0,0 +1,389 @@
+			key:      "healthcheck-client.key",
+		},
+	}
+
+	c, ok := lookup[distribution]
+	if !ok {
+		return nil, "", errors.Errorf("distribution %s not supported", distribution)
+	}
+
+	return []string{
+		"--cacert", c.hostPath + "/" + c.ca,
+		"--cert", c.hostPath + "/" + c.cert,
+		"--key", c.hostPath + "/" + c.key,
+		"--write-out", "json",
+		"--endpoints", "https://127.0.0.1:2379", // always use localhost
+	}, c.hostPath, nil
+}
+
+// getSupportedDistro returns the distro that etcd collector can run on
+// either due to the distro has static etcd pod (kurl by kubeadm) or
+// the distro has etcd running as a process (k0s, embedded-cluster)
+func (c *etcdDebug) getSupportedDistro() (string, error) {
+	// extract distro logic from analyzer.ParseNodesForProviders
+	// pkg/analyze/distribution.go
+	// we can't import analyzer because of circular dependency
+	// TODO: may refactor this to a common package
+
+	nodes, err := c.client.CoreV1().Nodes().List(c.context, metav1.ListOptions{})
+	if err != nil {
+		return "", errors.Wrap(err, "failed to list nodes")
+	}
+
+	for _, node := range nodes.Items {
+		for k, v := range node.ObjectMeta.Labels {
+			if k == "kurl.sh/cluster" && v == "true" {
+				return "kurl", nil
+			}
+			if k == "node.k0sproject.io/role" {
+				return "k0s", nil
+			}
+			if k == "kots.io/embedded-cluster-role" {
+				return "embedded-cluster", nil
+			}
+		}
+	}
+
+	return "", errors.New("current k8s distribution does not support etcd collector")
+}
+
+func (c *etcdDebug) getOrCreateEtcdPod() error {
+	// if ephemeral, create a etcd client pod to exec into
+	// the pod will use hostNetwork: true to access the etcd server
+	if c.ephemeral {
+		err := c.createEtcdPod()
+		if err != nil {
+			return errors.Wrap(err, "failed to create etcd pod")
+		}
+		return nil
+	}
+	// if not ephemeral, find the static etcd pod to exec into
+	// get the first etcd pod in the cluster with label "component=etcd" in all namespaces
+	label := "component=etcd"
+	pods, err := c.client.CoreV1().Pods("").List(context.Background(), metav1.ListOptions{
+		LabelSelector: label,
+	})
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to list etcd pods with label %s", label))
+	}
+	if len(pods.Items) == 0 {
+		return errors.New("no static etcd pod found")
+	}
+
+	klog.V(2).Infof("found etcd pod %s in namespace %s", pods.Items[0].Name, pods.Items[0].Namespace)
+	c.pod = &pods.Items[0]
+	return nil
+}
+
+// createEtcdPod creates a etcd client pod to exec into
+func (c *etcdDebug) createEtcdPod() error {
+	namespace := "default"
+	labels := map[string]string{
+		"troubleshoot-role": "etcd-collector",
+	}
+	spec := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "etcd-collector-",
+			Namespace:    namespace,
+			Labels:       labels,
+		},
+		Spec: corev1.PodSpec{
+			HostNetwork: true,
+			Containers: []corev1.Container{
+				{
+					Name:    "etcd-client",
+					Image:   "quay.io/coreos/etcd:latest",
```

Comment:
I think this ought to be passed in as input. In airgap EC environments, we would bundle an etcd image and then pass in the image name to the collector i.e the image would be something like `registry.namespace/etcd:tag`

Reply:
that's a great idea! I've made the requested changes. please let me know how it goes/